llvm/lib/Target/RISCV/RISCVInstrInfoZb.td

```diff
@@ -235,10 +235,7 @@ def SimmShiftRightBy3XForm : SDNodeXForm<imm, [{
 }]>;
 
 // Pattern to exclude simm12 immediates from matching.
-def non_imm12 : PatLeaf<(XLenVT GPR:$a), [{
-  auto *C = dyn_cast<ConstantSDNode>(N);
-  return !C || !isInt<12>(C->getSExtValue());
-}]>;
+def non_imm12 : ComplexPattern<XLenVT, 1, "selectNonImm12", [], [], 0>;
```

Comment:
Can you add `let GISelPredicateCode` here to do this without changing to a complex pattern?

Reply:
After some digging, I think the answer is no and I'm sad about it.

Reply:
Correct, using `GISelPredicateCode` was my first approach until I found that the `GlobalISelEmitter` TG backend doesn't pick that up for leaf nodes.

Reply:
Would add GISelPredicateCode to something like this work https://github.com/topperc/llvm-project/commit/01205c136d77f682063a1218790f7cf0b3760fdc

Reply:
> Would add GISelPredicateCode to something like this work [topperc@01205c1](https://github.com/topperc/llvm-project/commit/01205c136d77f682063a1218790f7cf0b3760fdc)

This works, though I'm a little concerned that this might create too many boilerplate code in the future, since there needs to be a Predicate TG record for every opcode that goes with non_imm12 (even we abstract the real predicate logics into a function or something). What do you think? 

Also, interestingly `GISelPredicateCode` doesn't dance well with `PredicateCodeUsesOperands`: it SEGFAULT llvm-tblgen in our case, despite the fact that there are tests for this exact combination (in `test/TableGen/GlobalISelEmitterCustomPredicate.td`).  I can't find an obvious fix for llvm-tblgen but writing a non-`PredicateCodeUsesOperands` predicate code also works.

Reply:
I used `PredicateCodeUsesOperands` so that I could know which operand wasn't the `shl` since add is commutable and tblgen will generate both patterns.

Reply:
> Would add GISelPredicateCode to something like this work [topperc@01205c1](https://github.com/topperc/llvm-project/commit/01205c136d77f682063a1218790f7cf0b3760fdc)

Done: it's no longer using ComplexPattern of `non_imm12` but PatFrag of `<op>_with_non_imm12` instead